noisy_simulator/src/instrument.rs

```diff
@@ -151,19 +135,32 @@ fn summed_kraus_operators(operations: &[Operation]) -> Result<Vec<SquareMatrix>,
     Ok(summed_kraus_operators)
 }
 
-/// Caches `vector_to_matrix_index` into a vector.
-fn cache_vector_to_matrix_index(operations: &[Operation]) -> Vec<(usize, usize)> {
+fn compute_choi_matrix(operations: &[Operation]) -> SquareMatrix {
     operations
-        .first()
+        .iter()
         .map(|op| {
-            if let Some(k) = op.kraus_operators().first() {
-                let num_elements = k.shape().0.pow(2);
-                (0..num_elements)
-                    .map(|idx| k.vector_to_matrix_index(idx))
-                    .collect()
-            } else {
-                Vec::default()
-            }
+            op.kraus_operators()
+                .iter()
+                .map(|k| {
+                    let vectorized_k = vectorize(k);
+                    &vectorized_k * &vectorized_k.adjoint()
+                })
+                .sum::<SquareMatrix>()
         })
-        .unwrap_or_default()
+        .sum()
+}
+
+/// Stacks the columns of `matrix` into a single column vector.
+///
+/// Perf transformation note: Typically vectorization stacks the
+/// rows of a matrix into a single column vector. But since we
+/// transposed all matrices until now, we stack the columns instead.
+///
+/// See `noisy_simulator/src/operation.rs/Operation::new` for more details.
+fn vectorize<T: nalgebra::Scalar + Copy>(matrix: &DMatrix<T>) -> DVector<T> {
+    let mut vectorized_matrix = DVector::<T>::from_vec(Vec::<T>::new());
+    for column in matrix.column_iter() {
+        vectorized_matrix.extend(column.iter().copied());
+    }
```

Comment:
Instead of extending, can we initialize the `vectorized_matrix` as we should know its final size before the loop